### DISCUSSION TITLE
x-ref not working when using expression generated name

### DISCUSSION THREAD

**Discussion by Briareos17** [alpinejs/alpine#3084] [7. Bugs (archived)] [answered]
When I trying to generate ref name with `:x-ref="refName('field_name', 1)"` alpine add separate `x-ref="field_name_1"` but when I trying to get ref from component, it returns undefined.

JS ex.:
```
document.addEventListener('alpine:init', () => {
        Alpine.data('form', () => ({
            fieldName(group, name, id) {
                return group + '[' + id + ']' + '[' + name + ']';
            },
            refName(name, id) {
                return name + '_' + id;
            },
            getRef(){
                console.log(this.$refs[this.refName('field_name', 1)]); // undefined
            }
        }))
    })
```

HTML ex:
```
<div x-data>
<input
    type="text"
    class="form-control"
    :name="fieldName('group_name','field_name', 1)"
    :x-ref="refName('field_name', 1)"
>
</div>
```

**Comment by SimoTod** [comment]
You can't dynamically generate x-ref in v3. It's in the documentation.

**Reply by Briareos17** [reply]
This is complete bullshit, I need to build list of repeated fields by rows, in each row I need to have access to sibling fields to make calculations and ref is the best for this but I need to use querySelector by my x-ref attributes instead of normal aproach...

**Reply by Tim-Wils** [reply, marked as answer]
I don't know your exact usecase, but you could most probably solve this with `x-id` https://alpinejs.dev/magics/id and/or `this.$root.querySelector('input')` or `this.$el`

**Reply by SimoTod** [reply]
> This is complete bullshit, I need to build list of repeated fields by rows, in each row I need to have access to sibling fields to make calculations and ref is the best for this but I need to use querySelector by my x-ref attributes instead of normal aproach...

I suggest you use an appropriate language when using a free tool that you didn't contribute to and it's likely to help you do your job to pay your bills. If you really need it, figure out how to update the library and send a PR. Have a good day.
